tests/run-build.sh

```diff
@@ -10,53 +10,81 @@ if test -e ci-skip ; then
     exit 0
 fi
 
-# Setup go environment on semaphore
-if [ -f /opt/change-go-version.sh ]; then
-    . /opt/change-go-version.sh
-    change-go-version 1.4
-fi
-
+SRC_CHANGES=1 # run functional tests by default
+DOC_CHANGES=1 # process docs by default
 RKT_STAGE1_USR_FROM="${1}"
 RKT_STAGE1_SYSTEMD_VER="${2}"
-
 BUILD_DIR="build-rkt-${RKT_STAGE1_USR_FROM}-${RKT_STAGE1_SYSTEMD_VER}"
 
+if [ "${CI-}" == true ] ; then
+	# https://semaphoreci.com/docs/available-environment-variables.html
+	if [ "${SEMAPHORE-}" == true ] ; then
+        # We might not need to run functional tests
+        SRC_CHANGES=$(git diff-tree --no-commit-id --name-only -r HEAD..origin/master | \
+            grep -cEv "^Documentation/")
+        # We might not need to process the docs
+        DOC_CHANGES=$(git diff-tree --no-commit-id --name-only -r HEAD..origin/master | \
+            grep -cE "^Documentation/")
+
+        # Setup go environment on semaphore
+        if [ -f /opt/change-go-version.sh ]; then
+            . /opt/change-go-version.sh
+            change-go-version 1.4
+        fi
+    fi
+fi
+
+if [ ${SRC_CHANGES} -eq 0 ] && [ ${DOC_CHANGES} -eq 0 ]; then
+    echo No changes detected.
+    exit 0
+fi
+
+# In case it wasn't cleaned up
+if [ -e "${BUILD_DIR}" ]; then
+    sudo rm -rf "${BUILD_DIR}"
+fi
+
 mkdir -p builds
 cd builds
 
-# Semaphore does not clean git subtrees between each build.
-sudo rm -rf "${BUILD_DIR}"
-
 git clone ../ "${BUILD_DIR}"
+pushd "${BUILD_DIR}"
+
+if [ ${SRC_CHANGES} -gt 0 ]; then 
+    echo Changes in sources detected. Running functional tests.
+    ./autogen.sh
+    case "${RKT_STAGE1_USR_FROM}" in
+        coreos|kvm)
+        ./configure --with-stage1="${RKT_STAGE1_USR_FROM}" \
+                --enable-functional-tests
+        ;;
+        host|usr-from-host)
+        ./configure --with-stage1=host \
+                --enable-functional-tests=auto
+        ;;
+        src)
+        ./configure --with-stage1="${RKT_STAGE1_USR_FROM}" \
+                --with-stage1-systemd-version="${RKT_STAGE1_SYSTEMD_VER}" \
+                --enable-functional-tests
+        ;;
+        *)
+        ./configure --with-stage1="${RKT_STAGE1_USR_FROM}"
+        ;;
+    esac
+
+    CORES=$(grep -c ^processor /proc/cpuinfo)
+    echo "Running make with ${CORES} threads"
+    make "-j${CORES}"
+    make check
+    make "-j${CORES}" clean
+fi
+if [ ${DOC_CHANGES} -gt 0 ]; then
+    :
+    # echo Changes in docs detected. Running functional tests.
```

Comment:
You probably meant "Runnning some doc checks" instead of "functional tests". Also, it might be good to extend the TODO below with what actually needs to be done (like check links validity).